### PR TITLE
feat(axis): provide splitLine.showMinLine and splitLine.showMaxLine. …

### LINKED ISSUE #16946: [Feature] xAxis splitLine showMinLine showMaxLine
### What problem does this feature solve?

When use splitLine in xAxis[type=time], minSplitLine and maxSplitLine can't be hidden.

image

### What does the proposed API look like?

like xAxisLabel [showMinLabel](https://echarts.apache.org/zh/option.html#xAxis.axisLabel.showMinLabel)

```ts
xAxis: {
  splitLine: {
     showMinLine: false,
     showMaxLine: true
  }
}
```

### PR DESCRIPTION
**Pull request by MeetzhDing**
## Brief Information

This pull request is in the type of:

- [ ] bug fixing
- [x] new feature
- [ ] others

### What does this PR do?

Resolves #16946

provide splitLine.showMinLine and splitLine.showMaxLine.

When using xAxis[type=time], it allow user hide first/last splitLine.

uplot style: 
uplot styles

### After: How is it fixed in this PR?
Like showMinLabel/showMaxLabel, it add new option `showMinLine` adn `showMaxLine`.

## Document Info

One of the following should be checked.

- [ ] This PR doesn't relate to document changes
- [x] The document should be updated later
- [ ] The document changes have been made in apache/echarts-doc#xxx

### CODE REVIEW

**Comment by pissang** [comment]
Please don't commit the files in the dist folder

File: src/coord/Axis.ts [hunk 1]
```diff
@@ -193,6 +193,13 @@ class Axis {
             this, ticksCoords, alignWithLabel, opt.clamp
         );
 
+        if (tickModel.get('showMinLine') === false) {
```
Comment (pissang) [inline src/coord/Axis.ts:196]: Modifying the result here will affect the display of labels and ticks lines. If only splitLine will be affected, it's better to do it in the render function https://github.com/apache/echarts/blob/cc0ebd2845781021cb05287fded7c94b26174619/src/component/axis/CartesianAxisView.ts#L141

Comment (MeetzhDing) [reply src/coord/Axis.ts]: done.

File: src/component/axis/CartesianAxisView.ts [hunk 2]
```diff
@@ -135,6 +135,13 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
             tickModel: splitLineModel
         });
 
+        if (splitLineModel.get('showMinLine') === false) {
```
Comment (pissang) [inline src/component/axis/CartesianAxisView.ts:138]: We usually use `!splitLineModel.get('showMinLine')` to check false value.

Comment (MeetzhDing) [reply src/component/axis/CartesianAxisView.ts]: d

**Comment by Ovilia** [comment]
I have some concern on this PR.

I think it may be a better idea to hide min/max line only if the first/last tick is not calculated by *nice tick* (which usually happens when the user deliberately sets a min/max value to the axis.

For example, in [this demo](https://echarts.apache.org/examples/zh/editor.html?c=line-function&code=GYVwdgxgLglg9mABKSAKAHgSkQbwFCKLqID0AvIgIwAMA3AYgE4CmUIjSAsgIZQAWAOgDOMMBmwAqRD34CIcIRkRSATIgDUVSdN6CRY4lIDMGxCu0BWOngC-eFNHhIA5szDNGvZgBFe3VNj4hAA2rIgAJn6IFADaALr0hMBwjIiooVCIMNGIALQqdFmIADwUBbRF6hTUApSBDISRUNwCAA4gQnyoMTAANMjgEKgwmHGYiYh2hCxsHBF-9HZwrbAIOUGI3GAwALa8TgBcyNzBQsy9DM6MMOFHG4RQy0cALNQXhCHMwFBHVu8f12cfB-iFe_0IACM4FBHjtftQGDZ_ugAILoGBCO4NRBgbg7ZhHADk6EJ4NIJEQO1ER3yb2xQlawRgUAAMqICbhsR9EJ04AB3I5QRggc5cj5M9wAZSgAE9QljuYrEPJgikiSxwoSxYQph9dUiGDK0RiFR9cfiiTKtTr_mdrsxMYgYtj7h9Za0OYSJcxSWLeXzJTKdlDgkdgCczmTCBAma1BcLRUqmtwjq53J4oD4_AFsbq4rZaEA), the min tick is `-20` and it is expected to be displayed because `-20` is a *nice tick*.

image


While in [another case](https://echarts.apache.org/examples/zh/editor.html?c=line-function&code=GYVwdgxgLglg9mABKSAKAHgSkQbwFCKLqID0AvIgIwAMA3AYgE4CmUIjSAsgIZQAWAOgDOMMBmwAqRD34CIcIRkRSATIgDUVSdN6CRY4lIDMGxCu0BWOngC-eFNHhIA5szDNGvZgBFe3VNj4hAA2rIgAJn6IFADaALr0hMBwjIiooVCIMNGIALQqdFmIADwUBbRF6hTUApSBDISRUNwCAA4gQnyoMTAANMjgEKgwmHGYiYh2hCxsHBF-9HZwrbAIOUGI3GAwALa8TgBcyNzBQsy9DM6MMOFHG4RQy0cALNQXhCHMwFBHVu8f12cfB-iFe_0IACM4FBHjtftQGDZ_ugAILoGBCO4NRBgbg7ZhHADk6EJ4MQO1ER3yKjJQlawRgUAAMqICbhsR9EJ04AB3I5QRggc4cj4M9wAZSgAE9QljOfLEPJgikiSxwoSRYQph9tUiGFK0Ri5R9cfiiVKNVr_mdrsxMYgYtj7h9pa02YSxcxSSLuTzxVKdlDgkdgCczmTCBAGa1-YLhQqmtwjq53J4oD4_AFsdq4rZaEA), the axis is set to have `min: -22`. In this case, the splitLine of `-22` is expected to be hidden if `showMinLine` is set to be `false`.

image


I think this logic may make more sense. @pissang What do you think?

**Comment by MeetzhDing** [comment]
@Ovilia @pissang 
how to judge whether it is `nice tick`?   Or is this any plan yet?

**Comment by Ovilia** [comment]
@MeetzhDing Sorry for the late reply. You can find the code [here](https://github.com/apache/echarts/blob/master/src/scale/Interval.ts#L218-L288).

**Comment by MeetzhDing** [comment]
@Ovilia showMinLine type is  ` boolean | 'auto'` now
```ts
    // null/undefined (true) | true | false | 'auto' (true when nick tick only)
    showMinLine?: boolean | 'auto',
    // null/undefined (true) | true | false | 'auto' (true when nick tick only)
    showMaxLine?: boolean | 'auto',
```

File: src/component/axis/CartesianAxisView.ts [hunk 3]
```diff
@@ -135,6 +139,37 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
             tickModel: splitLineModel
         });
 
+        if (ticksCoords.length > 2 && (axis.scale as any).getInterval instanceof Function) {
```
Comment (Ovilia) [inline src/component/axis/CartesianAxisView.ts:142]: Create a new variable for `axis.scale as IntervalScale`

Comment (Ovilia) [reply src/component/axis/CartesianAxisView.ts:142]: `instanceof Function` can be omitted. If `getInterval` is not null, we can expect it to be a function.

File: src/component/axis/CartesianAxisView.ts [hunk 4]
```diff
@@ -135,6 +139,37 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
...
+            const interval = (axis.scale as IntervalScale).getInterval() || null;
+            const precision = (axis.scale as IntervalScale).getIntervalPrecision instanceof Function
+                ? (axis.scale as IntervalScale).getIntervalPrecision()
+                : null;
+
+            const showMinLine = splitLineModel.get('showMinLine');
+            if (showMinLine === false
+                || (showMinLine === 'auto'
+                    && roundNumber(
```
Comment (Ovilia) [inline src/component/axis/CartesianAxisView.ts:151]: Using `roundNumber` to check if it is added as the *nice* extent is not accurate.

File: src/coord/axisCommonTypes.ts [hunk 5]
```diff
@@ -244,6 +244,10 @@ interface MinorTickOption {
 
 interface SplitLineOption {
     show?: boolean,
+    // null/undefined (true) | true | false | 'auto' (true when nick tick only)
+    showMinLine?: boolean | 'auto',
+    // null/undefined (true) | true | false | 'auto' (true when nick tick only)
```
Comment (Ovilia) [inline src/coord/axisCommonTypes.ts:249]: `null/undefined` is not necessary and they should not have a default value defined here. Instead, the default value should be defined in `axisDefault.ts`.

**Review by Ovilia — CHANGES_REQUESTED** [review]

**Comment by MeetzhDing** [comment]
@Ovilia  please check pr again

**Comment by MeetzhDing** [comment]
@Ovilia please check pr again

File: src/component/axis/CartesianAxisView.ts [hunk 6]
```diff
@@ -135,6 +136,30 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
             tickModel: splitLineModel
         });
 
+        const axisScale = axis.scale as IntervalScale;
+        if (ticksCoords.length > 2 && axisScale.getInterval) {
+            const interval = axisScale.getInterval() || null;
```
Comment (Ovilia) [inline src/component/axis/CartesianAxisView.ts:141]: What is `|| null` for?

File: src/component/axis/CartesianAxisView.ts [hunk 7]
```diff
@@ -135,6 +136,30 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
...
+
+            const showMinLine = splitLineModel.get('showMinLine');
+            if (showMinLine === false
+                || (showMinLine === 'auto'
+                    && ticksCoords[1].tickValue - ticksCoords[0].tickValue < interval
```
Comment (Ovilia) [inline src/component/axis/CartesianAxisView.ts:146]: Have you tested the case when `axis.inverse` is `true`? `ticksCoords[1].tickValue` may be smaller than `ticksCoords[0].tickValue` in this case.

File: src/component/axis/CartesianAxisView.ts [hunk 8]
```diff
@@ -135,6 +136,30 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
             tickModel: splitLineModel
         });
 
+        const axisScale = axis.scale as IntervalScale;
+        if (ticksCoords.length > 2 && axisScale.getInterval) {
```
Comment (Ovilia) [inline src/component/axis/CartesianAxisView.ts:140]: What if `ticksCoords.length` is `2` but the data range is not equal to the ticks' values?

File: test/axis-interval.html [hunk 9]
```diff
@@ -534,8 +540,68 @@ <h1>Test xAxis.axisTick.interval</h1>
 
 
 
+    <script>
+
+        require([
```
Comment (Ovilia) [inline test/axis-interval.html:545]: Please add more test cases to test boundary situations. You may start with the cases in my above questions.

Comment (Ovilia) [inline test/axis-interval.html:545]: Please add more test cases to test boundary situations. You may start with the cases in my above questions.

File: src/component/axis/CartesianAxisView.ts [hunk 10]
```diff
@@ -135,6 +136,30 @@ const axisElementBuilders: Record<typeof selfBuilderAttrs[number], AxisElementBu
...
+            const interval = axisScale.getInterval() || null;
+
+            const showMinLine = splitLineModel.get('showMinLine');
+            if (showMinLine === false
```
Comment (Ovilia) [inline src/component/axis/CartesianAxisView.ts:144]: `!showMinLine` should be better.

**Review by Ovilia — CHANGES_REQUESTED** [review]
Thanks for your contribution! I think the PR is close to the solution, only a few small tips.
Also, if you don't have some special reason, please use regular push instead of force push.

File: src/component/axis/CartesianAxisView.ts (hunk 8 above)
Comment (MeetzhDing) [reply src/component/axis/CartesianAxisView.ts:140]: When ticksCoords length is 2,  the splitLine is not expected to be hidden

**Comment by MeetzhDing** [comment]
image
image

@Ovilia please check again

File: src/component/axis/CartesianAxisView.ts (hunk 8 above)
Comment (Ovilia) [reply src/component/axis/CartesianAxisView.ts:140]: Why? For example, when data range is 11~19 and ticksCoords is [10, 20], I believe there should be no split line, right?

**Review by Ovilia — COMMENTED** [review]
I've run visual tests and LGTM. @plainheart would you like to double-check this PR?

**Comment by plainheart** [comment]
Superseded by #20114
